core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java

```diff
@@ -0,0 +1,440 @@
+                    .forEach(entry -> {
+                        Expression value = entry.getValue();
+                        if (entry.getValue() instanceof SymbolReference) {
+                            Symbol skewedSymbol = Symbol.from(value);
+                            if (context.containsKey(Symbol.from(value)) && !context.containsKey(entry.getKey())) {
+                                context.put(entry.getKey(), context.get(skewedSymbol));
+                            }
+                        }
+                    });
+
+            // Remove symbols that are dropped in the projection
+            context.keySet().removeAll(symbolsToRemove);
+
+            return replaceChildren(node, ImmutableList.of(child));
+        }
+
+        @Override
+        protected PlanNode visitPlan(PlanNode node, Map<Symbol, SkewedContext> context)
+        {
+            // For all other nodes, don't let the context pass through.
+            List<PlanNode> children = node.getSources().stream()
+                    .map(child -> child.accept(this, new HashMap<>()))
+                    .collect(Collectors.toList());
+            return replaceChildren(node, children);
+        }
+
+        @Override
+        public PlanNode visitTableScan(TableScanNode node, Map<Symbol, SkewedContext> context)
+        {
+            final String qualifiedTableName = metadata.getTableMetadata(session, node.getTable()).getQualifiedName().toString();
+
+            if (this.skewedTables.containsKey(qualifiedTableName)) {
+                SkewJoinConfig skewConfig = this.skewedTables.get(qualifiedTableName);
+                Optional<Pair<Symbol, ColumnMetadata>> symbolAndMetadata = node.getAssignments().entrySet().stream()
+                        .map(entry -> new Pair<>(entry.getKey(), metadata.getColumnMetadata(session, node.getTable(), entry.getValue())))
+                        .filter(entry -> entry.getB().getName().equals(skewConfig.getColumnName()))
+                        .findFirst();
+                symbolAndMetadata.ifPresent(m -> {
+                    context.putIfAbsent(m.getA(), new SkewedContext(skewConfig.values, toSqlType(m.getB().getType())));
+                });
+            }
+
+            return node;
+        }
+    }
+
+    static class SkewJoinConfig
+    {
+        /*
+         * Parses and validates the user supplied session data.
+         * */
+
+        final String tableName;
+        final String columnName;
+        final List<String> values;
+
+        private static final int TABLE_INDEX = 0;
+        private static final int COLUMN_INDEX = 1;
+        private static final int VALUE_INDEX = 2;
+        private static final int MIN_SIZE = 3;
+
+        private SkewJoinConfig(List<String> rawMetadata)
+        {
+            tableName = rawMetadata.get(TABLE_INDEX);
+            columnName = rawMetadata.get(COLUMN_INDEX);
+            values = ImmutableList.copyOf(rawMetadata.subList(VALUE_INDEX, rawMetadata.size()));
```

Comment:
This should be done in the `getMetadata` method below, and the object should be constructed with the constituent parts directly.